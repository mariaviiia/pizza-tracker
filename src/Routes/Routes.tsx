import App from "../App";
import Order from "../pages/0rder";
import Sales from "../pages/Sales";
import UserInfo from "../pages/UserInfo";
import NotFoundPage from "../pages/NotFoundPage";
import {
  Route,
  BrowserRouter as Router,
  Switch,
  Redirect,
} from "react-router-dom";
import { createContext, useState, useEffect } from "react";
import { IngredientModel } from "../components/common/Ingredient/Ingredient";

export interface PizzaModel {
  name: string;
  ingredients: IngredientModel[];
}

export interface OrderModel {
  userName: string;
  phone: string;
  date: string;
  pizzas: PizzaModel[];
  total: number;
}

export const PizzasContext = createContext<{
  pizzas: PizzaModel[];
  setPizzas: (newPizzas: PizzaModel[]) => void;
  orders: OrderModel[];
  setOrders: (newOrders: OrderModel[]) => void;
}>([] as any);

const Routes = () => {
  const [pizzas, setPizzas] = useState<PizzaModel[]>([]),
    [orders, setOrders] = useState<OrderModel[]>([]);

  useEffect(() => {
    const localOrders = JSON.parse(
      localStorage.getItem("pizzaTrackerOrders") || "[]"
    );
    setOrders(localOrders);
  }, []);

  return (
    <PizzasContext.Provider value={{ pizzas, setPizzas, orders, setOrders }}>
      <Router>
        <div className="page">
          <Switch>
            <Route exact path="/" component={App} />
            <Route exact path="/Order" component={Order} />
            <Route exact path="/UserInfo" component={UserInfo} />
            <Route exact path="/sales" component={Sales} />
            <Route path="/404" component={NotFoundPage} />
            <Redirect to="/404" />
          </Switch>
        </div>
      </Router>
    </PizzasContext.Provider>
  );
};
export default Routes;
