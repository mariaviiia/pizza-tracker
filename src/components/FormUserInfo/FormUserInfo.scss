.user-info {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  flex-direction: column;
  &__form {
    width: 200px;
    display: flex;
    flex-direction: column;
    span {
      font-size: 12px;
      font-family: cursive;
      margin-bottom: 0.5em;
    }
    input {
      margin-bottom: 1.5em;
      border: 1px solid gray;
      border-radius: 5px;
      padding: 5px;
      font-size: 10px;
      width: 100%;
    }
  }
  &__details {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 150px;
    margin-bottom: 2.5em;
    &--title {
      font-size: 16px;
      font-family: cursive;
      letter-spacing: 0.5px;
      margin: 1em 0;
    }
    &--description {
      width: 100%;
      display: flex;
      justify-content: space-between;
      font-size: 12px;
      span {
        margin-right: 1px;
      }
    }
  }
}
