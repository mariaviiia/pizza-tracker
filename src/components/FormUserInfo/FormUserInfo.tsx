import { useContext, useState } from "react";
import { Link } from "react-router-dom";
import { PizzasContext } from "../../Routes/Routes";
import Button from "../common/Button/Button";
import getPizzaCost from "../../Utilities/getPizzaCost";
import "./FormUserInfo.scss";

const FormUserInfo = () => {
  const { pizzas, orders, setOrders } = useContext(PizzasContext),
    [userName, setUserName] = useState<string>(""),
    [phone, setPhone] = useState<string>(""),
    [date, setDate] = useState<string>(""),
    isFinishButtonEnabled = () => userName && phone && date,
    save = () => {
      const total = pizzas.reduce(
          (count, { ingredients }) => count + getPizzaCost(ingredients),
          0
        ),
        newOrders = [
          ...orders,
          {
            userName,
            phone,
            date,
            pizzas,
            total,
          },
        ];

      localStorage.setItem("pizzaTrackerOrders", JSON.stringify(newOrders));
      setOrders(newOrders);
    };

  return (
    <div className="user-info">
      <div className="user-info__form">
        <span>Nombre*</span>
        <input
          type="text"
          onChange={({ target }) => {
            setUserName(target.value);
          }}
          value={userName}
        />
        <span>Telefono*</span>
        <input
          type="number"
          onChange={({ target }) => {
            setPhone(target.value);
          }}
          value={phone}
        />
        <span>Fecha*</span>
        <input
          type="date"
          onChange={({ target }) => {
            setDate(target.value);
          }}
          value={date}
        />
      </div>
      <div className="user-info__details">
        <span className="user-info__details--title">order details</span>
        {pizzas.map(({ name, ingredients }) => {
          return (
            <div key={name} className="user-info__details--description">
              <span>{name}:</span> {ingredients.length} ingredients.
            </div>
          );
        })}
      </div>
      <Link to="/sales">
        <Button
          onClick={save}
          disabled={!isFinishButtonEnabled()}
          buttonLabel="Finish"
        />
      </Link>
    </div>
  );
};

export default FormUserInfo;
