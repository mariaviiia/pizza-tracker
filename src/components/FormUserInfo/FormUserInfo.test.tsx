import FormUserInfo from "./FormUserInfo";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { PizzasContext } from "../../Routes/Routes";

describe("PizzaMaker component", () => {
  const MockPizzas = {
    pizzas: [],
    setPizzas: () => undefined,
    orders: [],
    setOrders: () => undefined,
  };
  it("renders correctly", () => {
    const { container } = render(
      <PizzasContext.Provider value={MockPizzas}>
        <Router>
          <FormUserInfo />
        </Router>
      </PizzasContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
