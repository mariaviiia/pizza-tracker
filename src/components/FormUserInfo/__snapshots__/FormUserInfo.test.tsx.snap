// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PizzaMaker component renders correctly 1`] = `
<div>
  <div
    class="user-info"
  >
    <div
      class="user-info__form"
    >
      <span>
        Nombre*
      </span>
      <input
        type="text"
        value=""
      />
      <span>
        Telefono*
      </span>
      <input
        type="number"
        value=""
      />
      <span>
        Fecha*
      </span>
      <input
        type="date"
        value=""
      />
    </div>
    <div
      class="user-info__details"
    >
      <span
        class="user-info__details--title"
      >
        order details
      </span>
    </div>
    <a
      href="/sales"
    >
      <button
        class="button button-primary button__size-md"
        disabled=""
      >
        Finish
      </button>
    </a>
  </div>
</div>
`;
